File: gmshParser.m
function msh = gmshParser(filename)

%% Reads a mesh in msh format, version 4.1
%
% Usage: 
% To define all variables m.LINES, M.TRIANGLES, etc 
% (Warning: This creates a very large structure. Do you really need it?)
%            m = load_gmsh4('a.msh')
%
% To define only certain variables (for example TETS and HEXS)
%            m = load_gmsh4('a.msh', [ 5 6])
%
% To define no variables (i.e. if you prefer to use m.ELE_INFOS(i,2))
%            m = load_gmsh4('a.msh', -1)
%            m = load_gmsh4('a.msh', [])
%
% Copyright (C) 2007  JP Moitinho de Almeida (moitinho@civil.ist.utl.pt)
% and  R Lorphevre(r(point)lorphevre(at)ulg(point)ac(point)be)
%
% based on load_gmsh.m supplied with gmsh-2.0
%
% Structure msh always has the following elements:
%
% msh.MIN, msh.MAX - Bounding box
% msh.nbNod - Number of nodes
% msh.nbElm - Total number of elements
% msh.nbType(i) - Number of elements of type i (i in 1:19)
% msh.POS(i,j) - j'th coordinate of node i (j in 1:3, i in 1:msh.nbNod)
% msh.ELE_INFOS(i,1) - id of element i (i in 1:msh.nbElm)
% msh.ELE_INFOS(i,2) - type of element i
% msh.ELE_INFOS(i,3) - number of tags for element i
% msh.ELE_INFOS(i,4) - Dimension (0D, 1D, 2D or 3D) of element i
% msh.ELE_TAGS(i,j) - Tags of element i (j in 1:msh.ELE_INFOS(i,3))
% msh.ELE_NODES(i,j) - Nodes of element i (j in 1:k, with
%                       k = msh.NODES_PER_TYPE_OF_ELEMENT(msh.ELE_INFOS(i,2)))
%
% These elements are created if requested:
%
% msh.nbLines - number of 2 node lines
% msh.LINES(i,1:2) - nodes of line i
% msh.LINES(i,3) - tag (WHICH ?????) of line i
%
% msh.nbTriangles - number of 2 node triangles
% msh.TRIANGLES(i,1:3) - nodes of triangle i
% msh.TRIANGLES(i,4) - tag (WHICH ?????) of triangle i
%
% Etc

% These definitions need to be updated when new element types are added to gmsh
%
% msh.Types{i}{1} Number of an element of type i
% msh.Types{i}{2} Dimension (0D, 1D, 2D or 3D) of element of type i
% msh.Types{i}{3} Name to add to the structure with elements of type i
% msh.Types{i}{4} Name to add to the structure with the number of elements of type i
%

msh.elementTypes = { ...
    { 2, 1, 'LIN_2', 'nbLines'}, ... % 1
    { 3,  2, 'TRI_3', 'nbTriangles'}, ...
    { 4,  2, 'QUA_4', 'nbQuads'}, ...  
    { 4,  3, 'TET_4', 'nbTets'}, ...
    { 8,  3, 'HEX_8', 'nbHexas'}, ... %5
    { 6,  3, 'PRI_6', 'nbPrisms'}, ...
    { 5,  3, 'PYR_5', 'nbPyramids'}, ...
    { 3,  1, 'LIN_3', 'nbLines3'}, ...
    { 6,  2, 'TRI_6', 'nbTriangles6'}, ...
    { 9,  2, 'QUA_9', 'nbQuads9'}, ... % 10
    { 10,  3, 'TET_10', 'nbTets10'}, ...
    { 27,  3, 'HEX_27', 'nbHexas27'}, ...
    { 18,  3, 'PRI_18', 'nbPrisms18'}, ...
    { 14,  3, 'PYR_14', 'nbPyramids14'}, ...
    { 1,  0, 'PNT', 'nbPoints'}, ... % 15
    { 8,  3, 'QUA_8', 'nbQuads8'}, ...
    { 20,  3, 'HEX_20', 'nbHexas20'}, ...
    { 15,  3, 'PRI_15', 'nbPrisms15'}, ...
    { 13,  3, 'PYR_13', 'nbPyramids13'}, ...
};
% TODO: Add the new element types of GMSH, see 
% https://gitlab.onelab.info/gmsh/gmsh/blob/master/Common/GmshDefines.h
% TODO: Add error handling for input parsing in sub functions

fid = fopen(filename, 'r');

tline = fgetl(fid);
if (feof(fid))
    fprintf('Empty file: %s \n',  filename);
    return;
end

%% Read the remaining lines and find the blocks with $Nodes and $Elements
while ~feof(fid)
    switch tline
        case '$MeshFormat' % Mandatory block
            msh.meshFormat = readMeshFormat(fid);
        case '$PhysicalNames'
            msh.physicalNames = readPhysicalNames(fid);
        case '$Entities'
            msh.entities = readEntities(fid);
        case '$Nodes'
            msh.nodes = readNodes(fid);
        case '$Elements'
            msh.elements = readElements(fid, msh.elementTypes);
    end
    tline = fgetl(fid);
end

fclose(fid);

end

function [meshFormat] = readMeshFormat(fid)
    data = fscanf(fid, '%f %d %d', 3);
    meshFormat.version = data(1);
    meshFormat.fileType = data(2);
    meshFormat.dataSize = data(3);
    
    if (meshFormat.version ~= 4.1)
        error('The GMSH file version %f, is not supported (only version 4.1 is) \n', form(1) );
    end

    if (meshFormat.fileType ~= 0)
        error('Sorry, this program can only read ASCII format');
    end
    
    flushBlock(fid, '$EndMeshFormat');
end

function [entities] = readEntities(fid)
    numData = fscanf(fid, '%d', 4);
    entities.numPoints = numData(1);
    entities.numCurves = numData(2);
    entities.numSurfaces = numData(3);
    entities.numVolumes = numData(4);
    
    for iPoint = 1:entities.numPoints
        data = fscanf(fid, '%d %f %f %f', 4);
        numPhysicalTags = fscanf(fid, '%d', 1);
        physicalTags = fscanf(fid, '%d', numPhysicalTags);
        entities.points(iPoint) = struct(...
            'pointTag', data(1), ...
            'X', data(2), ...
            'Y', data(3), ...
            'Z', data(4), ...
            'numPhysicalTags', numPhysicalTags, ...
            'physicalTags', physicalTags ...
        );
    end
    
    for iCurve = 1:entities.numCurves
        data = fscanf(fid, '%d %f %f %f %f %f %f', 7);
        numPhysicalTags = fscanf(fid, '%d', 1);
        physicalTags = fscanf(fid, '%d', numPhysicalTags);
        numBoundingPoints = fscanf(fid, '%d', 1);
        pointTags = fscanf(fid, '%d', numBoundingPoints);
        entities.curves(iCurve) = struct(...
            'curveTag', data(1), ...
            'minX', data(2), ...
            'minY', data(3), ...
            'minZ', data(4), ...
            'maxX', data(5), ...
            'maxY', data(6), ...
            'maxZ', data(7), ...
            'numPhysicalTags', numPhysicalTags, ...
            'physicalTags', physicalTags, ...
            'numBoundingPoints', numBoundingPoints, ...
            'pointTags', pointTags ...
        );
    end
    
    for iSurface = 1:entities.numSurfaces
        data = fscanf(fid, '%d %f %f %f %f %f %f', 7);
        numPhysicalTags = fscanf(fid, '%d', 1);
        physicalTags = fscanf(fid, '%d', numPhysicalTags);
        numBoundingCurves = fscanf(fid, '%d', 1);
        curveTags = fscanf(fid, '%d', numBoundingCurves);
        entities.surfaces(iSurface) = struct(...
            'surfaceTag', data(1), ...
            'minX', data(2), ...
            'minY', data(3), ...
            'minZ', data(4), ...
            'maxX', data(5), ...
            'maxY', data(6), ...
            'maxZ', data(7), ...
            'numPhysicalTags', numPhysicalTags, ...
            'physicalTags', physicalTags, ...
            'numBoundingCurves', numBoundingCurves, ...
            'curveTags', curveTags ...
        );
    end
    
    for iVolume = 1:entities.numVolumes
        data = fscanf(fid, '%d %f %f %f %f %f %f', 7);
        numPhysicalTags = fscanf(fid, '%d', 1);
        physicalTags = fscanf(fid, '%d', numPhysicalTags);
        numBoundingSurfaces = fscanf(fid, '%d', 1);
        surfaceTags = fscanf(fid, '%d', numBoundingSurfaces);
        entities.volumes(iVolume) = struct(...
            'volumeTag', data(1), ...
            'minX', data(2), ...
            'minY', data(3), ...
            'minZ', data(4), ...
            'maxX', data(5), ...
            'maxY', data(6), ...
            'maxZ', data(7), ...
            'numPhysicalTags', numPhysicalTags, ...
            'physicalTags', physicalTags, ...
            'numBoundingSurfaces', numBoundingSurfaces, ...
            'surfaceTags', surfaceTags ...
        );
    end
    
    flushBlock(fid, '$EndEntities');
end

function [pNames] = readPhysicalNames(fid)
    pNames.numPhysicalNames = fscanf(fid, '%d', 1);
    % Use textscan instead of fscanf to parse string
    data = textscan(fid, '%d %d %s', [3 pNames.numPhysicalNames])';
    % Fill structure array backwards to hopefully allocate all memory at
    % once
    for ips = pNames.numPhysicalNames:-1:1
        name = data{3}{ips};
        % trim the quotes from the start and the end of the string
        name = name(2:end-1);
        
        pNames.physicalNames(ips) = struct(...
            'dimension', data{1}(ips), ...
            'physicalTag', data{2}(ips), ...
            'name', name ...
        );
    end

    flushBlock(fid, '$EndPhysicalNames');
end

function [nodes] = readNodes(fid)
    metaData = fscanf(fid, '%d', 4); % <numEntityBlocks> <numNodes> <minNodeTag> <maxNodeTag>
    nodes.numEntityBlocks = metaData(1);
    nodes.numNodes = metaData(2);
    nodes.minNodeTag = metaData(3);
    nodes.maxNodeTag = metaData(4);
    
    for iEntityBlock = 1:nodes.numEntityBlocks
        entityData = fscanf(fid, '%d', 4); % <entityDim> <entityTag> <parametric> <numNodesInBlock>
        if entityData(3)
            disp('The parametric option is not yet supported by the parser');
            break;
        end
        nodeTags = fscanf(fid, '%g', entityData(4));
        pos = fscanf(fid, '%g', [3, entityData(4)])'; % transpose pos as fscanf inserts in column order
        u = []; v = []; w = [];  
        
        nodes.entityBlocks(iEntityBlock) = struct(...
            'entityDim', entityData(1), ...
            'entityTag', entityData(2), ...
            'parametric', entityData(3), ...
            'numNodesInBlock', entityData(4), ...
            'nodeTags', nodeTags, ...
            'x', pos(:, 1), ...
            'y', pos(:, 2), ...
            'z', pos(:, 3), ...
            'u', u, ...
            'v', v, ...
            'w', w ...
        );
    end
    
    flushBlock(fid, '$EndNodes');
end

function [elements] = readElements(fid, elementTypes)
    metaData = fscanf(fid, '%f', 4); % <numEntityBlocks> <numElements> <minElementTag> <maxElementTag>
    
    elements.numEntityBlocks = metaData(1);
    elements.numElements= metaData(2);
    elements.minElementTag = metaData(3);
    elements.maxElementTag = metaData(4);
    
    for iEntity = 1:elements.numEntityBlocks
        entityData = fscanf(fid, '%d', 4); % <entityDim> <entityTag> <elementType> <numNodesInBlock>
        
        numNodes = elementTypes{entityData(3)}{1};
         % transpose elements as fscanf inserts in column order
        elementDataMatrix = fscanf(fid, '%f', [1+numNodes entityData(4)])'; % <elementTag> <nodeTag> ... 
        
        for iElem = size(elementDataMatrix, 1):-1:1
            elems(iElem) = struct( ...
                'elementTag', elementDataMatrix(iElem, 1), ...
                'nodeTags', elementDataMatrix(iElem, 2:end) ...
            );
        end
        elements.entityBlocks(iEntity) = struct( ...
            'entityDim', entityData(1), ...
            'entityTag', entityData(2), ...
            'elementType', entityData(3), ...
            'numElementsInBlock', entityData(4), ...
            'elements', elems ...
        );
    end
    
    flushBlock(fid, '$EndElements');
end

function flushBlock(fid, endTag)
    fgetl(fid);     % flush pointer to new line
    tline = fgetl(fid);     % this should be $EndElements
    if (~strcmp(tline, endTag))
        fprintf('Syntax error (no %s) \n',  endTag);
    end
end

File: julia/GmshParser.jl
using DelimitedFiles

export ElementType, parse_mesh, Gmsh

@enum ElementType::Int begin
    LIN_2 = 1
    TRI_3 = 2
    QUA_4 = 3
    TET_4 = 4
    HEX_5 = 5
    PRI_6 = 6
    PYR_5 = 7
    LIN_3 = 8
    TRI_6 = 9
    QUA_9 = 10
    TET_10 = 11
    HEX_27 = 12
    PRI_18 = 13
    PYR_14 = 14
    PNT = 15
end

struct MeshFormat
    version::AbstractFloat
    filetype::Integer
    datasize::Integer
end

struct PhysicalName
    dimension::Integer
    physical_tag::Integer
    name::AbstractString
end

struct Point
    tag::Integer
    x::Float64
    y::Float64
    z::Float64
    physical_tags::AbstractArray{Int}
end

struct Curve
    tag::Integer
    min_x::Float64
    min_y::Float64
    min_z::Float64
    max_x::Float64
    max_y::Float64
    max_z::Float64
    physical_tags::AbstractArray{Int}
    point_tags::AbstractArray{Int}
end

struct Surface
    tag::Integer
    min_x::Float64
    min_y::Float64
    min_z::Float64
    max_x::Float64
    max_y::Float64
    max_z::Float64
    physical_tags::AbstractArray{Int}
    curve_tags::AbstractArray{Int}
end

struct Volume
    tag::Integer
    min_x::Float64
    min_y::Float64
    min_z::Float64
    max_x::Float64
    max_y::Float64
    max_z::Float64
    physical_tags::AbstractArray{Int}
    surface_tags::AbstractArray{Int}
end

struct Node
    node_tag::Integer
    x::Float64
    y::Float64
    z::Float64
end

struct NodeBlock
    entity_dim::Integer
    entity_tag::Integer
    parametric::Bool
    nodes::AbstractArray{Node, 1}
end

struct Element
    element_tag::Integer
    node_tags::AbstractArray{Int}
end

struct ElementBlock
    entity_dim::Integer
    entity_tag::Integer
    element_type::ElementType
    elements::AbstractArray{Element, 1}

    ElementBlock(a, b, c, d) = new(a, b, ElementType(c), d)
end

mutable struct Gmsh
    format::MeshFormat
    physical_names::AbstractArray{PhysicalName, 1}

    points::AbstractArray{Point, 1}
    curves::AbstractArray{Curve, 1}
    surfaces::AbstractArray{Surface, 1}
    volumes::AbstractArray{Volume, 1}

    node_blocks::AbstractArray{NodeBlock, 1}
    element_blocks::AbstractArray{ElementBlock, 1}

    Gmsh() = new()
end



function parse_mesh(filename::AbstractString)
    #@info filename
    file = open(filename, "r")
    gmsh = Gmsh()
    while !eof(file)
        blockname = readline(file)
        if blockname == "\$MeshFormat"
            gmsh.format = parse_meshformat(file)
        elseif blockname == "\$PhysicalNames"
            gmsh.physical_names = parse_physical_names(file)
        elseif blockname == "\$Entities"
            entities = parse_entities(file)
            gmsh.points = entities[1]
            gmsh.curves = entities[2]
            gmsh.surfaces = entities[3]
            gmsh.volumes = entities[4]
        elseif blockname == "\$Nodes"
            gmsh.node_blocks = parse_nodes(file)
        elseif blockname == "\$Elements"
            gmsh.element_blocks = parse_elements(file)
        end
    end
    close(file)

    return gmsh
end

function parse_meshformat(file::IOStream)
    meshformat_data = parse_vector(file)
    @info "GMSH-file version $(meshformat_data[1])"
    return MeshFormat(meshformat_data...)
end

function parse_physical_names(file::IOStream)
    num_physical_names = parse_vector(file; type=Int)[1]
    @info "Found $num_physical_names physical names:"
    physical_names = Array{PhysicalName, 1}(undef, num_physical_names)
    for i = 1:num_physical_names
        line = readline(file)
        line_vector = split(line, ' ')
        numericals = map(x -> parse(Float64, x), line_vector[1:2])
        name = trim(line_vector[3])
        @info "\t $name"
        physical_names[i] = PhysicalName(numericals..., name)
    end
    return physical_names
end

function parse_entities(file::IOStream)
    # <num-points num-curves num-surfaces num-volumes>
    num_data = parse_vector(file; type=Int)
    @info "Found $(num_data[1]) points"
    @info "Found $(num_data[2]) curves"
    @info "Found $(num_data[3]) surfaces"
    @info "Found $(num_data[4]) volumes"
    points = Array{Point, 1}(undef, num_data[1])
    curves = Array{Curve, 1}(undef, num_data[2])
    surfaces = Array{Surface, 1}(undef, num_data[3])
    volumes = Array{Volume, 1}(undef, num_data[4])

    for i = 1:num_data[1]
        data = parse_vector(file)
        num_physical_tags = convert(Int, data[5])
        points[i] = Point(data[1:4]..., convert.(Int, data[6:5 + num_physical_tags]))
    end

    function fill_entities!(entities, type::DataType, index::Integer, data)
        num_physical_tags = convert(Int, data[8])
        num_bounding_tags = convert(Int, data[8 + num_physical_tags + 1])

        if num_physical_tags > 0
            # Bounding tags exist
            physical_tags = convert.(Int, data[9:9+num_physical_tags-1])
        else
            physical_tags = Array{Int,1}(undef, 0)
        end
        if num_bounding_tags > 0
            # Bounding tags exist
            bounding_tags = convert.(Int, data[8+num_physical_tags+2:end])
        else
            bounding_tags = Array{Int,1}(undef, 0)
        end
        entities[index] = type(data[1:7]..., physical_tags, bounding_tags)
    end

    for i = 1:num_data[2]
        fill_entities!(curves, Curve, i, parse_vector(file))
    end
    for i = 1:num_data[3]
        fill_entities!(surfaces, Surface, i, parse_vector(file))
    end
    for i = 1:num_data[4]
        fill_entities!(volumes, Volume, i, parse_vector(file))
    end

    return (points, curves, surfaces, volumes)
end

function parse_nodes(file::IOStream)
    # <numEntityBlocks> <numNodes> <minNodeTag> <maxNodeTag>
    num_data = parse_vector(file; type=Int)
    @info "Detected $(num_data[2]) nodes"
    node_blocks = Array{NodeBlock, 1}(undef, num_data[1])

    for i = 1:num_data[1]
        # <entityDim> <entityTag> <parametric> <numNodesInBlock>
        entity_data = parse_vector(file; type=Int)
        if Bool(entity_data[3])
            @error "The parametric option is not yet supported by the parser"
        end
        node_tags = Array{Int, 1}(undef, entity_data[4])
        for j = 1:entity_data[4]
            node_tags[j] = parse_vector(file; type=Int)[1]
        end
        nodes = Array{Node, 1}(undef, entity_data[4])
        for j = 1:entity_data[4]
            pos = parse_vector(file)
            nodes[j] = Node(node_tags[j], pos...)
        end
        node_blocks[i] = NodeBlock(entity_data[1:3]..., nodes)
    end

    return node_blocks
end

function parse_elements(file::IOStream)
    # <numEntityBlocks> <numElements> <minElementTag> <maxElementTag>
    num_data = parse_vector(file; type=Int)
    @info "Detected $(num_data[2]) elements"
    element_blocks = Array{ElementBlock, 1}(undef, num_data[1])

    for i = 1:num_data[1]
        # <entityDim> <entityTag> <elementType> <numElementsInBlock>
        entity_data = parse_vector(file; type=Int)
        elements = Array{Element, 1}(undef, entity_data[4])

        for j = 1:entity_data[4]
            elem_data = parse_vector(file; type=Int)
            elements[j] = Element(elem_data[1], elem_data[2:end])
        end

        element_blocks[i] = ElementBlock(entity_data[1:3]..., elements)
    end

    return element_blocks
end
# Helper functions

function parse_vector(file::IOStream; delim::AbstractChar = ' ',
        type::DataType = Float64)
    line = readline(file)
    line_vector = split(line, delim)
    line_vector = filter(x -> x != "", line_vector)
    float_vector = map(x -> parse(type, x), line_vector)
end

function trim(str)
    str[2:end-1]
end

function get_physical_element_blocks(gmsh::Gmsh, physical_name::AbstractString)
    found_groups = filter(pn -> pn.name == physical_name, gmsh.physical_names)
    if isempty(found_groups)
        @error "No group with the physical name $physical_name found"
    else
        @debug "Found $physical_name"
    end

    blocks = Array{ElementBlock, 1}(undef, 0)
    found_element_blocks = Array{ElementBlock, 1}(undef, 0)
    for group = found_groups
        if group.dimension == 0
            elementary_entities = gmsh.points
        elseif group.dimension == 1
            elementary_entities = gmsh.curves
        elseif group.dimension == 2
            elementary_entities = gmsh.surfaces
        elseif group.dimension == 3
            elementary_entities = gmsh.volumes
        else
            @error "The dimension parameter of the physical group is invalid"
        end

        # Find which entities that are included in the physical group
        found_entities = filter(e -> any(e.physical_tags .== group.physical_tag), elementary_entities)

        # Extract the element blocks which are included in the found entities
        correct_entity(e) = any(e.entity_tag .== (p->p.tag).(found_entities)) &&
            e.entity_dim == group.dimension

        append!(found_element_blocks, filter(correct_entity, gmsh.element_blocks))
    end

    return found_element_blocks
end

function get_global_coordinates(gmsh::Gmsh)
    number_nodes = get_number_nodes(gmsh)
    node_coordinates = zeros(number_nodes, 3)
    for block = gmsh.node_blocks, node = block.nodes
        node_coordinates[node.node_tag, 1] = node.x
        node_coordinates[node.node_tag, 2] = node.y
        node_coordinates[node.node_tag, 3] = node.z
    end

    return node_coordinates
end

function get_number_nodes(gmsh::Gmsh)
    number_nodes = 0
    for block = gmsh.node_blocks
        number_nodes += length(block.nodes)
    end
    return number_nodes
end

function get_elements(gmsh::Gmsh, element_type::ElementType)
    blocks = filter(block -> block.element_type == element_type, gmsh.element_blocks)
    Enod = Array{Array{Float64}, 1}(undef, 0)

    function concat_elements(element_block)
        element_tags = [e.element_tag for e in element_block.elements]
        node_tags = transpose(hcat([e.node_tags for e in element_block.elements]...))
        return hcat(element_tags, node_tags)
    end

    Enod = vcat(map(concat_elements, blocks)...)
end
